### ISSUE TITLE
Refine only current hunk

### ISSUE THREAD

**Issue by tarsius** [magit/magit#1787]
*As suggested by @uu1101 in #1581:*

Thinking a bit more about this, I wonder why should section highlight and hunk refinement go hand-in-hand.  That is, when I set `magit-diff-refine-hunk` to `t`, I want the hunk I am looking at to be refined, not on all hunks descendant of currently selected section.  The patch below refines hunks only when they are the current section; the disk churn is gone and the result is equally usable.

It seems that when `magit-diff-refine-hunk` is `all`, the refinement happens before, when washing, so this does not seem to interfere.

What do you think?

~~~{.patch}
diff --git i/magit-diff.el w/magit-diff.el
index f23f6d9..81c1dbd 100644
--- i/magit-diff.el
+++ w/magit-diff.el
@@ -1596,7 +1596,8 @@ of SECTION including SECTION and all of them are highlighted."
                                  'magit-diff-whitespace-warning)))))
 
 (defun magit-diff-refine-hunk (hunk)
-  (unless (magit-section-refined hunk)
+  (unless (or (magit-section-refined hunk)
+               (not (eq (magit-current-section) hunk)))
     (setf (magit-section-refined hunk) t)
     (save-excursion
       (goto-char (magit-section-start hunk))
}

**Comment by tarsius** [comment]
That's actually how it used to work. When I changed it to refine hunks that are children of the current section instead of only the current hunk, I did not realize that this could have a negative impact on performance.

For the time being I have added a new variable `magit-diff-refine-hunk-related` which defaults to nil. It's undocumented and not exposed as an option, but at least it allows users who do not agree with this change to revert it.

The variables and especially the toggle command need an overhaul anyway, this just adds another complication that I'll have to take care of at some point. But now is not a good time for it.
